### PR TITLE
Track expression dependencies

### PR DESCRIPTION
**Pull request by TimSylvester**
When building expression trees, we assign and capture what categories of input they depend on, like feature data and zoom level.

When evaluating whether to update a layer, zoom changes only trigger an update if zoom is used in an expression somewhere within the style.

---

I discovered that the benchmarking I was doing was thrown off by some local changes lingering in my workspace, so I'm re-evaluating...

deps:

```
Average frame encoding time: 2.650 ms
Average frame rendering time: 6.487 ms

Average frame encoding time: 2.606 ms
Average frame rendering time: 6.392 ms

Average frame encoding time: 2.579 ms
Average frame rendering time: 6.360 ms

Average frame encoding time: 2.673 ms
Average frame rendering time: 6.564 ms

Average frame encoding time: 2.606 ms
Average frame rendering time: 6.536 ms

Average frame encoding time: 2.592 ms
Average frame rendering time: 6.444 ms
```

Mean encoding: 2.6177

main:
```
Average frame encoding time: 2.664 ms
Average frame rendering time: 6.47 ms

Average frame encoding time: 2.711 ms
Average frame rendering time: 6.451 ms

Average frame encoding time: 2.639 ms
Average frame rendering time: 6.372 ms

Average frame encoding time: 2.624 ms
Average frame rendering time: 6.472 ms

Average frame encoding time: 2.715 ms
Average frame rendering time: 6.536 ms

Average frame encoding time: 2.662 ms
Average frame rendering time: 6.427 ms
```

Mean encoding: 2.6692

Difference: 1.95%

### CODE REVIEW

**Review by sjg-wdw — APPROVED** [review]

**Review by stefankarschti — APPROVED** [review]

File: src/mbgl/map/map_impl.cpp [hunk 1]
```diff
@@ -84,6 +86,33 @@ void Map::Impl::onStyleLoaded() {
         annotationManager.onStyleLoaded();
     }
     observer.onDidFinishLoadingStyle();
+
+#if !defined(NDEBUG)
```
Comment (louwers) [inline src/mbgl/map/map_impl.cpp:90]: Should be extracted in a separate function.

File: src/mbgl/style/properties.hpp [hunk 2]
```diff
@@ -298,6 +339,23 @@ class Properties {
                                other.template get<Ps>().value))...});
             return result;
         }
+
+        Dependency getDependencies() const noexcept {
```
Comment (louwers) [inline src/mbgl/style/properties.hpp:343]: `std::make_tuple` is not `noexcept`.

**Review by louwers — CHANGES_REQUESTED** [review]
Could you add tests?

File: src/mbgl/style/properties.hpp (hunk 2 above)
Comment (TimSylvester) [reply src/mbgl/style/properties.hpp:343]: This is evaluated at compile time, though I forgot to put `constexpr` on it.  But I can do it more manually to eliminate any ambiguity.

image

Comment (louwers) [reply src/mbgl/style/properties.hpp:349]: Ah OK, but this is one compiler.

I would err on the side of safety in general when it comes to `noexcept`.

**Comment by TimSylvester** [comment]
> Could you add tests?

Done.  I also extended some test cases because the modifications to existing code with low coverage made CodeCov mad, so overall coverage should be up a bit.

**Comment by TimSylvester** [comment]
> Could you add tests?

Done.  I also expanded some existing test cases, as modifications to some areas with low coverage brought the average way down.

**Review by louwers — APPROVED** [review]

**Pull request merged**